src/rust/engine/src/externs/fs.rs

```diff
@@ -426,21 +415,62 @@ impl<'source> FromPyObject<'source> for PyPathGlobs {
   }
 }
 
-#[pyfunction]
-fn match_path_globs(
-  py_path_globs: PyPathGlobs,
-  paths: Vec<String>,
-  py: Python,
-) -> PyResult<Vec<String>> {
-  py.allow_threads(|| {
-    let path_globs = py_path_globs.parse()?;
-    Ok(
-      paths
-        .into_iter()
-        .filter(|p| path_globs.matches(Path::new(p)))
-        .collect(),
-    )
-  })
+// -----------------------------------------------------------------------------
+// FilespecMatcher
+// -----------------------------------------------------------------------------
+
+#[pyclass(name = "FilespecMatcher")]
+#[derive(Debug)]
+pub struct PyFilespecMatcher(FilespecMatcher);
+
+#[pymethods]
+impl PyFilespecMatcher {
+  #[new]
+  fn __new__(includes: Vec<String>, excludes: Vec<String>) -> PyResult<Self> {
+    let matcher = FilespecMatcher::new(includes, excludes).map_err(PyValueError::new_err)?;
+    Ok(Self(matcher))
+  }
+
+  fn __hash__(&self) -> u64 {
+    let mut s = DefaultHasher::new();
+    self.0.include_globs().hash(&mut s);
+    self.0.exclude_globs().hash(&mut s);
+    s.finish()
+  }
+
+  fn __repr__(&self) -> String {
+    let includes = self
+      .0
+      .include_globs()
+      .iter()
+      .map(|pattern| format!("{pattern}"))
+      .join(", ");
+    let excludes = self.0.exclude_globs().join(", ");
+    format!("FilespecMatcher(includes=['{includes}'], excludes=[{excludes}])",)
+  }
+
+  fn __richcmp__(&self, other: &PyFilespecMatcher, op: CompareOp, py: Python) -> PyObject {
+    match op {
+      CompareOp::Eq => (self.0.include_globs() == other.0.include_globs()
+        && self.0.exclude_globs() == other.0.exclude_globs())
+      .into_py(py),
+      CompareOp::Ne => (self.0.include_globs() != other.0.include_globs()
+        || self.0.exclude_globs() != other.0.exclude_globs())
+      .into_py(py),
+      _ => py.NotImplemented(),
+    }
+  }
+
+  fn matches(&self, paths: Vec<String>, py: Python) -> PyResult<Vec<String>> {
+    py.allow_threads(|| {
```

Comment:
@stuhood has suggested it may be more efficient to remove this

Reply:
Would need to do some microbenchmarking, yea.

The two cases are:
1. taking `paths: Vec<&str>`, which would avoid copying the inputs, but would require us to hold the GIL, or
2. keeping as is

You could maybe temporarily add a loop around the `match` calls in `src/python/pants/source/filespec_test.py` `assert_rule_matches`, and then compare that way. It's not a completely realistic use case, but it would be straightforward. I expect that it's not worth doing anything more involved though.